### PR TITLE
WIP: initial avoid loop

### PR DESCRIPTION
**Pull request by TheReturnOfJan**

### CODE REVIEW

File: src/Oracle.sol [hunk 1]
```diff
@@ -6,19 +7,20 @@ import "./Machine.template.sol";
 
 interface IOracle {
 
-  // questionKey is initialStateHash          
+  // questionKey is initialStateHash
   struct Question {
     uint askTime;
     uint timeout;
     bytes32[] answerKeys;
+    uint amountOfAnswers;
```
Comment (TheReturnOfJan) [inline src/Oracle.sol:15]: Maybe rename to numberOfUnfalsifiedAnswers.

File: src/Oracle.sol [hunk 2]
```diff
@@ -155,13 +159,23 @@ contract Oracle is IOracle {
...
-    require(_answerExists(answerKey), "The answer trying to be falsified does not exist");
+    require(_answerExists(answerKey), "The answer trying to be falsified does not exist or was already falsified.");
     require(msg.sender == court, "Only court can falsify answers");
 
-    answer.falsified = true;
-    payable(prosecutor).call.value(STAKE_SIZE)("");
+    Question storage question = questions[answer.questionKey];
+    bytes32 questionKey = answer.questionKey;
+
+    if (answer.position == question.answerKeys.length - 1) {
```
Comment (TheReturnOfJan) [inline src/Oracle.sol:168]: Extract this stuff into an internal _removeAnswer.

File: src/Oracle.sol [hunk 3]
```diff
@@ -175,14 +189,22 @@ contract Oracle is IOracle {
     require(_questionExists(answer.questionKey) && _answerExists(answerKey), "Question and answer must exists.");
     require(now >= question.askTime + question.timeout, "Answering is still in progress.");
     require(Machine.imageHash(image) == answerKey, "Image hash does not match answerKey.");
-    require(!answer.falsified, "This answer was falsified");
+
+    // Maybe here we should add check to ensure that all wrong answers were already falsified and if no just increase timeout
+    if (question.amountOfAnswers > 1) {
```
Comment (TheReturnOfJan) [inline src/Oracle.sol:194]: Change into a require.

File: src/Oracle.sol [hunk 4]
```diff
@@ -202,26 +224,23 @@ contract Oracle is IOracle {
 
     function(bytes32) external callback = question.failCallback;
 
-    _questionCleanup(questionKey);
-    
+    for (uint i = 0; i < question.answerKeys.length; i ++) {
+      if (question.answerKeys[i] > 0) {
+        bytes32 answerKey = question.answerKeys[i];
```
Comment (TheReturnOfJan) [inline src/Oracle.sol:229]: Do we to delete the entry in answerKeys array aswell here 🤔 ?

Comment (roleengineer) [reply src/Oracle.sol]: yeap

File: src/Oracle.sol [hunk 5]
```diff
@@ -243,4 +242,38 @@ contract Oracle is IOracle {
...
+
+  function _removeAnswer (
+    bytes32 answerKey
+  ) internal
+  {
+    Answer storage answer = answers[answerKey];
+    Question storage question = questions[answer.questionKey];
+
+    if (answer.indexInQuestionArray == question.answerKeys.length - 1) {
+      question.answerKeys.pop();
```
Comment (roleengineer) [inline src/Oracle.sol:254]: do we need this?

**Comment by TheReturnOfJan** [comment]
I approve.

**Pull request merged**